compiler/noirc_frontend/src/elaborator/comptime.rs

```diff
@@ -329,15 +329,14 @@ impl<'context> Elaborator<'context> {
                 push_arg(Value::TraitDefinition(trait_id));
             } else {
                 let (expr_id, expr_type) = interpreter.elaborator.elaborate_expression(arg);
-                push_arg(interpreter.evaluate(expr_id)?);
-
                 if let Err(UnificationError) = expr_type.unify(param_type) {
                     return Err(InterpreterError::TypeMismatch {
                         expected: param_type.clone(),
                         actual: expr_type,
                         location: arg_location,
                     });
                 }
+                push_arg(interpreter.evaluate(expr_id)?);
```

Comment:
To make sure I understand: in the `foo` example `expr_id` and `expr_type` corresponded to `32`, and was assigned as `Field` by default, but `unify` looks at the definition of `foo`, `expr_type` is sort of _unbound_, and gets refined to `u32`, right?

Reply:
Exactly! I don't exactly understand how `unify` does that because none of the arguments are `&mut`, but maybe it does that through interior mutability (which I don't know well yet).

Reply:
It indeed works through interior mutability on Types, specifically the TypeVariable struct which internally holds a `Shared<T>` which is a wrapper for a `Rc<RefCell<T>>`. The `RefCell<T>` enables shared mutability on a T, while the `Rc` makes it easily cloneable.